src/V3Randomize.cpp

```diff
@@ -464,7 +583,74 @@ class RandomizeMarkVisitor final : public VNVisitor {
+        if (randObject && nodep->varp()
+            && nodep->varp()->rand().isRandomizable()) {  // Process global constraints
+            if (m_classp && m_classp->user1() == IS_RANDOMIZED) {
+                m_classp->user1(IS_RANDOMIZED_GLOBAL);
+            }
+            // Mark the entire nested chain as participating in global constraints
+            // For foo.in.val, this marks: foo, foo.in, and foo.in.val
+            if (VN_IS(nodep->fromp(), VarRef) || VN_IS(nodep->fromp(), MemberSel)) {
+                markNestedGlobalConstrainedRecurse(nodep->fromp());
+            }
+
+            // Global constraint processing algorithm:
+            // 1. Detect globally constrained object variables in randomized classes
+            // 2. Clone constraint trees from the constrained object's class
+            // 3. Rename cloned constraints with object prefix (obj.var format)
+            // 4. Insert cloned constraints into current class for solver processing
+            // 5. Use basic randomization for non-constrained variables to avoid recursion
+
+            // Extract and validate components early to avoid repeated type checks
+            AstVarRef* const varRefp = VN_CAST(nodep->fromp(), VarRef);
+            if (!varRefp || !varRefp->varp()) return;
+
+            const AstClassRefDType* const classRefp
+                = VN_CAST(varRefp->dtypep()->skipRefp(), ClassRefDType);
+            UASSERT(classRefp && classRefp->classp(),
+                    "Global constraint variable must have valid class type");
+
+            if (nodep->user1() && varRefp->varp()->globalConstrained()) {
+                AstClass* gConsClass = classRefp->classp();
+                AstVar* const objVar = varRefp->varp();  // The object variable (e.g., obj1, obj2)
+
+                // Process per-variable (object instance), not per-class
+                // This allows multiple objects of the same class (e.g., obj1 and obj2 of type Sub)
+                if (m_processedVars.find(objVar) == m_processedVars.end()) {
+
+                    // Clone constraints from the top-level class (e.g., Level1 for obj_a)
+                    gConsClass->foreachMember(
+                        [&](AstClass* const classp, AstConstraint* const constrp) {
+                            UASSERT(constrp, "foreachMember should only pass valid constraints");
+                            AstConstraint* const cloneConstrp = constrp->cloneTree(false);
+                            // Name manipulation
+                            nameManipulation(varRefp, cloneConstrp);
+                            m_clonedConstraints.push_back(cloneConstrp);
+                        });
+
+                    // Recursively clone constraints from nested members (e.g., l2, l3, l4)
+                    cloneNestedConstraints(varRefp, gConsClass);
+
+                    // Mark this variable (object instance) as processed
+                    m_processedVars.insert(objVar);
+                }
```

Comment:
Suggest move this chunk to a new function as the visit's function is larger than a screen full.